src/snowflake/connector/connection.py

```diff
@@ -1049,6 +1058,47 @@ def _write_params_to_byte_rows(
             raise BindUploadError from exc
         return res
 
+    def _get_snowflake_type_and_binding(
+        self,
+        cursor: Optional["SnowflakeCursor"],
+        v: Union[Tuple[str, Any], Any],
+    ) -> TypeAndBinding:
+        if isinstance(v, tuple):
+            if len(v) != 2:
+                Error.errorhandler_wrapper(
+                    self,
+                    cursor,
+                    ProgrammingError,
+                    {
+                        "msg": "Binding parameters must be a list "
+                        "where one element is a single value or "
+                        "a pair of Snowflake datatype and a value",
+                        "errno": ER_FAILED_PROCESSING_QMARK,
+                    },
+                )
+            return TypeAndBinding(
+                v[0], self.converter.to_snowflake_bindings(v[0], v[1])
+            )
+        else:
+            snowflake_type = self.converter.snowflake_type(v)
+            if snowflake_type is None:
+                Error.errorhandler_wrapper(
+                    self,
+                    cursor,
+                    ProgrammingError,
+                    {
+                        "msg": "Python data type [{}] cannot be "
+                        "automatically mapped to Snowflake data "
+                        "type. Specify the snowflake data type "
+                        "explicitly.".format(v.__class__.__name__.lower()),
+                        "errno": ER_NOT_IMPLICITY_SNOWFLAKE_DATATYPE,
+                    },
+                )
+            return TypeAndBinding(
+                snowflake_type,
+                self.converter.to_snowflake_bindings(snowflake_type, v),
+            )
```

Comment:
I have the tiniest nitpick:
The return statement can be moved out from the if-else clauses this way:
```suggestion
            snowflake_type, v = v
        else:
            snowflake_type = self.converter.snowflake_type(v)
            if snowflake_type is None:
                Error.errorhandler_wrapper(
                    self,
                    cursor,
                    ProgrammingError,
                    {
                        "msg": "Python data type [{}] cannot be "
                        "automatically mapped to Snowflake data "
                        "type. Specify the snowflake data type "
                        "explicitly.".format(v.__class__.__name__.lower()),
                        "errno": ER_NOT_IMPLICITY_SNOWFLAKE_DATATYPE,
                    },
                )
        return TypeAndBinding(
            snowflake_type,
            self.converter.to_snowflake_bindings(snowflake_type, v),
        )
```